### PR TITLE
Feat(plugins): Add AVD String Formatter for later use in custom descriptions

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Feat(plugins): Add AVD String Formatter for later use in custom descriptions

## Proposed changes

- New string formatter, reusing parts of Python string.Formatter class.
- Support for upper
- Support for optional fields
- Support for prefix / suffix - useful with optional fields
- Protection against accessing internal objects.
- Easily extensible to add IP addressing or other things.

## How to test

Added pytest

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: python-avd/pyavd/_utils/format_string.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            format_spec = org_format_spec
+
+            # output the literal text
+            if literal_text:
+                result.append(literal_text)
+
+            # if there's a field, output it
+            if field_name is not None:
+                # this is some markup, find the object and do
+                #  the formatting
```
Comment (gmuloc) [inline python-avd/pyavd/_utils/format_string.py]:
```suggestion
                # this is some markup, find the object and do the formatting
```

File: python-avd/tests/pyavd/utils/test_format_string.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    pytest.param("{<foo  >bar  }", ("Ethernet1",), {}, "foo  Ethernet1bar  ", id="positional_field_with_prefix_and_suffix_existing_arg"),
+    pytest.param("{?<foo>bar}", (), {}, "", id="positional_optional_field_with_prefix_and_suffix_missing_arg"),
+    pytest.param("{?< f o o > b a r }", (None,), {}, "", id="positional_optional_field_with_prefix_and_suffix_none_arg"),
+    pytest.param("{?< f o o > b a r }", ("Ethernet1",), {}, " f o o Ethernet1 b a r ", id="positional_optional_field_with_prefix_and_suffix_existing_arg"),
+    # positional fields with prefix and suffix and upper
+    pytest.param("{<foo  >bar  !u}", ("Ethernet1",), {}, "foo  ETHERNET1bar  ", id="positional_field_with_prefix_and_suffix_existing_arg_and_upper"),
+    pytest.param("{?<foo>bar!u}", (), {}, "", id="positional_optional_field_with_prefix_and_suffix_missing_arg_and_upper"),
+    pytest.param("{?< f o o > b a r !u}", (None,), {}, "", id="positional_optional_field_with_prefix_and_suffix_none_arg_and_upper"),
+    pytest.param(
+        "{?< f o o > b a r !u}", ("Ethernet1",), {}, " f o o ETHERNET1 b a r ", id="positional_optional_field_with_prefix_and_suffix_existing_arg_and_upper"
```
Comment (gmuloc) [inline python-avd/tests/pyavd/utils/test_format_string.py:85]: does it work if I want `! / < / >` in my suffix / prefix?

File: python-avd/tests/pyavd/utils/test_format_string.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    ),
+    # positional fields with prefix and suffix
+    pytest.param("{<foo  >bar  }", ("Ethernet1",), {}, "foo  Ethernet1bar  ", id="positional_field_with_prefix_and_suffix_existing_arg"),
+    pytest.param("{?<foo>bar}", (), {}, "", id="positional_optional_field_with_prefix_and_suffix_missing_arg"),
+    pytest.param("{?< f o o > b a r }", (None,), {}, "", id="positional_optional_field_with_prefix_and_suffix_none_arg"),
+    pytest.param("{?< f o o > b a r }", ("Ethernet1",), {}, " f o o Ethernet1 b a r ", id="positional_optional_field_with_prefix_and_suffix_existing_arg"),
+    # positional fields with prefix and suffix and upper
+    pytest.param("{<foo  >bar  !u}", ("Ethernet1",), {}, "foo  ETHERNET1bar  ", id="positional_field_with_prefix_and_suffix_existing_arg_and_upper"),
+    pytest.param("{?<foo>bar!u}", (), {}, "", id="positional_optional_field_with_prefix_and_suffix_missing_arg_and_upper"),
+    pytest.param("{?< f o o > b a r !u}", (None,), {}, "", id="positional_optional_field_with_prefix_and_suffix_none_arg_and_upper"),
```
Comment (gmuloc) [inline python-avd/tests/pyavd/utils/test_format_string.py:83]: the code should support digits like `"{0}{1}{0}".format("blah", "2")` would render `blah2blah` maybe lets add a test for this?

File: python-avd/pyavd/_utils/format_string.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    Custom string formatter class to provide extra protection from malicious format strings and support for prefixes and suffixes per field.
+
+    The regular Python syntax is "{" [field_name] ["!" conversion] [":" format_spec] "}"
+    This class supports "{" [field_name] ["?"] ["<" prefix] [">" suffix] ["!" conversion] [":" format_spec] "}"
+
+    where
+        ? ::= The literal ? signals that the field is optional and will not be printed if the value is missing or None.
+        prefix ::= string including spaces which will be inserted before the field value. Most useful in combination with ?
+        suffix ::= string including spaces which will be inserted after the field value. Most useful in combination with ?
```
Comment (gmuloc) [inline python-avd/pyavd/_utils/format_string.py]: cool things!

File: python-avd/pyavd/_utils/format_string.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        conversion ::= "!u" for "upper()" (The regular Python conversions "!r" | "!s" | "!a" have been removed).
```
Comment (gmuloc) [inline python-avd/pyavd/_utils/format_string.py]: is this for security purposes?

**Review by gmuloc — COMMENTED** [review]
nice PR!

File: python-avd/pyavd/_utils/format_string.py (hunk 5 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_utils/format_string.py]: well !s is the default for format strings. We do not need/want ascii() or repr().

File: python-avd/tests/pyavd/utils/test_format_string.py (hunk 3 above)
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/utils/test_format_string.py:83]: This is regular format string behavior. I am not changing that.

File: python-avd/tests/pyavd/utils/test_format_string.py (hunk 2 above)
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/utils/test_format_string.py:85]: it does not work. Either we can implement some form of escaping which is still fragile for the builtin things like ! and :, or we just call it out that !, :, < and > are not supported in prefix/suffix?

File: python-avd/tests/pyavd/utils/test_format_string.py (hunk 3 above)
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/utils/test_format_string.py:83]: added tests anyway

File: python-avd/tests/pyavd/utils/test_format_string.py (hunk 2 above)
Comment (ClausHolbechArista) [reply python-avd/tests/pyavd/utils/test_format_string.py:85]: Added restriction to doc string.

**Review by gmuloc — APPROVED** [review]

**Pull request merged**